### PR TITLE
Enable account registration

### LINKED ISSUE #1987: Activate user registration with confirmation email
The Jhipster template provides the possibility for user registration with confirmation emails.

This should be activated  in Artemis based on a new setting in the yml configuration. Only if user registration is enabled in this setting, it will be shown to users at the start page (e.g. a link below the login screen)

It should be possible to define multiple regular expressions (in the yml configuration) for the email that the users have to use for the registration.

Ideally the email template for the confirmation email is configurable (either as a config file that can be changed on the server by super admins, e.g. as the privacy statement or even better as part of the Artemis GUI in the system administration settings) to include specific information (such as the university name, etc.)

### PR DESCRIPTION
**Pull request by krusche**
This PR enables the registration functionality that is offered by default by the jhipster template.
By default, it is disabled. Admins can enable it in the yml file.
They can also specify a certain email regex pattern for specific emails (e.g. @tum.de).

Fixes #1987 

### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I added multiple integration tests (Spring) related to the features (with a high test coverage)
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I added multiple integration tests (Jest) related to the features (with a high test coverage)
- [x] Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context
Allow Artemis admins to activate the default Jhipster registration functionality.

### Description
additional client side improvements:
* eliminate lots of duplicated code in quiz statistics, fix a couple of issues with respect to the charts display
* adapt new jhipster code for components in the account module
* use a constant for roles
* remove sugar as it was not needed

### Missing
- [x] Improve error handling when the registration does not work, currently the error shown to the user is generic (Will be resolved in a follow up PR)
- [x] Email pattern display and validation on client side (the one that can specified by the admin in yml)
- [x] Testing all the different cases extensively 

### Steps for Testing on **Test Server** 2
1. Make sure you are not logged in
2. Click on Register and follow the steps (email -> activation -> login)
3. Check that the user is also created on Gitlab
4. Click on Passwort Reset and follow the steps
5. Check that the changed password was also updated on Gitlab after reset
6. Try out password change (in the user menu in the upper right corner)
7. Try out settings with name changes (in the user menu in the upper right corner)
8. Please also test quiz statistics and check everything is working properly (I removed lots of duplicated code)

### Test Coverage
* Server: 
    * AccountResource: 78%
    * UserService: 81%
* Client:
    * main/webapp/app/account/activate: 81 %
    * main/webapp/app/account/password: 80 %
    * main/webapp/app/account/password-reset/finish: 96 %
    * main/webapp/app/account/password-reset/init: 87 %
    * main/webapp/app/account/register: 83 %
    * main/webapp/app/account/settings: 94 %

### Screenshots

image

image

### CODE REVIEW

**Review by kloessst — CHANGES_REQUESTED** [review]
**Registration functionality:**
What worked:
- Registration and activation worked fine for me.
- Resetting the password via e-mail worked
- Resetting the password via the Artemis password menu worked
- Renaming worked

What didn't work so well:
- I had to wait quite some time to get the reset e-mails (maybe provider related?)
- GitLab required me to choose a new password after the reset. Not sure if this is expected
unexpectedAfterReset

**Quizzes:**
Not sure if these problems are related to this PR
- Could not create a quiz when importing questions from other quizzes.
quizCreationError

- Diagram labels for the short answer quiz statistics don't match or I don't know how to read the diagram 😄 
saQuizStatistics

**Comment by sleiss** [comment]
I can login with my new account before activating it using the email, is this intended?

File: docs/admin/registration.rst
Comment (sleiss) [inline docs/admin/registration.rst:13]: This line seems erroneous, as we don't expect a `.` in the domain, but allow any value.
This would e.g. allow the domain `inAtum.de` to register users.
I believe this can be fixed be escaping the dot.
```suggestion
                allowed-email-pattern: ([a-zA-Z0-9_\-\.\+]+)@(tum\.de)|(in\.tum\.de)|(mytum\.de)
```

**Comment by anditurdiu** [comment]
I was not able to register with a new account: 
image
At first I got an error 405: 
```
{
    "type":"https://www.jhipster.tech/problem/problem-with-message",
    "title":"Method Not Allowed",
    "status":405,"detail":
    "Request method 'GET' not supported",
    "path":"/api/register",
    "message":"error.http.405"
}
```
After hitting reload and trying to register again it says:
```
{
    "entityName":"userManagement",
    "errorKey":"emailexists",
    "type":"https://www.jhipster.tech/problem/email-already-used",
    "title":"Email is already in use!",
    "status":400,
    "message":"error.emailexists",
    "params":"userManagement"
}
```
I used the only @tum email address I have, andi.turdiu@tum.de not sure if this is expected or not

File: docs/admin/registration.rst
Comment (sleiss) [inline docs/admin/registration.rst:13]: Actually, there is another flaw as it seems:
```suggestion
                allowed-email-pattern: ([a-zA-Z0-9_\-\.\+]+)@(tum\.de|in\.tum\.de|mytum\.de)
```

Comment (krusche) [reply docs/admin/registration.rst:13]: I am not a regex expert 😅

**Comment by krusche** [comment]
> I was not able to register with a new account:
> image
> At first I got an error 405:
> 
> ```
> {
>     "type":"https://www.jhipster.tech/problem/problem-with-message",
>     "title":"Method Not Allowed",
>     "status":405,"detail":
>     "Request method 'GET' not supported",
>     "path":"/api/register",
>     "message":"error.http.405"
> }
> ```
> 
> After hitting reload and trying to register again it says:
> 
> ```
> {
>     "entityName":"userManagement",
>     "errorKey":"emailexists",
>     "type":"https://www.jhipster.tech/problem/email-already-used",
>     "title":"Email is already in use!",
>     "status":400,
>     "message":"error.emailexists",
>     "params":"userManagement"
> }
> ```
> 
> I used the only @TUM email address I have, [andi.turdiu@tum.de](mailto:andi.turdiu@tum.de) not sure if this is expected or not

You can use andi.turdiu+test@tum.de or anything else after the `+` and before the `@`

**Comment by krusche** [comment]
@kloessst not sure why it took longer, could be related to the email provider or the email setup that I have chosen for TS2.
The Gitlab issue with changing the password is something that we cannot solve on our own. There is a long standing issue in the Gitlab development itself. But I think you can simply replace the `old` password with the same `new` password

File: docs/admin/registration.rst
Comment (krusche) [reply docs/admin/registration.rst:13]: I think I found the 2nd error already, but only replaced it my local (not committed) application.yml file
Thanks for noticing it 👍

Comment (stefanwaldhauser) [reply docs/admin/registration.rst:13]: You can test it on https://regexr.com/ btw

Comment (stefanwaldhauser) [reply docs/admin/registration.rst:13]: I think it needs to be `([a-zA-Z0-9_\-\.\+]+)@((tum\.de)|(in\.tum\.de)|(mytum\.de))` otherwise

**Comment by anditurdiu** [comment]
Account creation, password reset works fine, it is also created and maintained as expected on gitlab. 
I can also confirm the issue @kloessst mentioned regarding importing quiz questions

File: src/main/webapp/app/account/password/password.component.html
Comment (sleiss) [inline src/main/webapp/app/account/password/password.component.html:5]: What is the reason for the `[`/`]` brackets? I find them quite uncommon in this context.

**Comment by stefanwaldhauser** [comment]
Could not create account
image

Error: https://sentry.io/share/issue/b634af6c508a49738742da7bc86a3b83/

**Comment by sleiss** [comment]
> Could not create account
> image
> 
> Error: https://sentry.io/share/issue/b634af6c508a49738742da7bc86a3b83/

I've adjusted the regex in the config of TS2, but not restarted. If you redeploy, it should work.

**Review by JonasPetry — COMMENTED** [review]
I get a 401 when I open the login page. Tested hard reloading and emptied the cache in Chrome. I've also tried Safari, but the error does not go away.

image

**Comment by sleiss** [comment]
> I get a 401 when I open the login page. Tested hard reloading and emptied the cache in Chrome. I've also tried Safari, but the error does not go away.
> 
> image

@JonasPetry 
That's actually intended (as long as it's only once): The client does not know if the user is logged in and thus tries to get the account details and determines if the user is logged in by either receiving the account details (user is logged in) or an error (user is not logged in).

File: src/main/java/de/tum/in/www1/artemis/web/rest/AccountResource.java
Comment (MaisiKoleni) [inline src/main/java/de/tum/in/www1/artemis/web/rest/AccountResource.java:104]: See #1659.
```suggestion
        if (registrationEnabled.isEmpty() || Boolean.FALSE.equals(registrationEnabled.get())) {
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/AccountResource.java
Comment (MaisiKoleni) [inline src/main/java/de/tum/in/www1/artemis/web/rest/AccountResource.java:78]: See #1659.
```suggestion
        if (registrationEnabled.isEmpty() || Boolean.FALSE.equals(registrationEnabled.get())) {
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/AccountResource.java
Comment (MaisiKoleni) [inline src/main/java/de/tum/in/www1/artemis/web/rest/AccountResource.java:163]: See #1659.
```suggestion
        if (registrationEnabled.isEmpty() || Boolean.FALSE.equals(registrationEnabled.get())) {
```

**Comment by stefanwaldhauser** [comment]
Tested:
Account Creation
Password Reset
Also works as expected in gitlab

File: src/main/java/de/tum/in/www1/artemis/web/rest/AccountResource.java
Comment (MaisiKoleni) [inline src/main/java/de/tum/in/www1/artemis/web/rest/AccountResource.java:186]: See #1659.
```suggestion
        if (registrationEnabled.isEmpty() || Boolean.FALSE.equals(registrationEnabled.get())) {
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/AccountResource.java
Comment (MaisiKoleni) [inline src/main/java/de/tum/in/www1/artemis/web/rest/AccountResource.java:202]: See #1659.
```suggestion
        if (registrationEnabled.isEmpty() || Boolean.FALSE.equals(registrationEnabled.get())) {
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/AccountResource.java
Comment (MaisiKoleni) [inline src/main/java/de/tum/in/www1/artemis/web/rest/AccountResource.java:225]: See #1659.
```suggestion
        if (registrationEnabled.isEmpty() || Boolean.FALSE.equals(registrationEnabled.get())) {
```

**Review by MaisiKoleni — CHANGES_REQUESTED** [review]
I have some suggestions regarding the `Boolean` values, see inline comments.

An alternative would also be
```java
        if (registrationEnabled.isEmpty() || !Boolean.TRUE.equals(registrationEnabled.get())) {
```

File: src/main/webapp/app/account/password/password-strength-bar.component.ts
Comment (derLalla) [inline src/main/webapp/app/account/password/password-strength-bar.component.ts:32]:
```suggestion
            return Boolean.True.equals(isMatchedFlag);
```

**Comment by krusche** [comment]
> **Quizzes:**
> Not sure if these problems are related to this PR
> 
> * Could not create a quiz when importing questions from other quizzes.

This issue should have been fixed in another branch

> * Diagram labels for the short answer quiz statistics don't match or I don't know how to read the diagram 😄

This could be an issue, I'll look into it

File: src/main/webapp/app/account/password/password.component.html
Comment (krusche) [reply src/main/webapp/app/account/password/password.component.html:5]: fixed

File: docs/admin/registration.rst
Comment (krusche) [reply docs/admin/registration.rst:13]: fixed

Comment (krusche) [reply docs/admin/registration.rst:13]: fixed

**Review by sleiss — DISMISSED** [review]
Tested account registration locally with Gitlab and on TS2, worked fine.
Was also able to configure the email provider and verified that account activation is required.

TODO for a follow-up: Show that the login failed due to a non-activated account.

File: src/main/resources/config/application-artemis.yml
Comment (MaisiKoleni) [inline src/main/resources/config/application-artemis.yml:26]: We should be able to do that as well:
```suggestion
            allowed-email-pattern: [a-zA-Z0-9_\-\.\+]+@(tum\.de|in\.tum\.de|mytum\.de)
```

Comment (sleiss) [reply src/main/resources/config/application-artemis.yml:26]: Checked it, also works like this.

File: src/main/webapp/app/account/register/register.component.html
Comment (MaisiKoleni) [inline src/main/webapp/app/account/register/register.component.html:12]: Is this still needed?
```suggestion
```

**Comment by krusche** [comment]
> Tested account registration locally with Gitlab and on TS2, worked fine.
> Was also able to configure the email provider and verified that account activation is required.
> 
> TODO for a follow-up: Show that the login failed due to a non-activated account.

Thanks. Improved error messages will be implemented in a follow-up

File: src/main/webapp/app/account/register/register.component.ts
Comment (MaisiKoleni) [inline src/main/webapp/app/account/register/register.component.ts:54]: Is this still needed?
```suggestion
```

File: src/main/resources/config/application-artemis.yml
Comment (krusche) [reply src/main/resources/config/application-artemis.yml:26]: thanks, this looks easier, I'll change it

**Review by MaisiKoleni — DISMISSED** [review]
Some commented-out test code is still left, if identified it correctly.

File: src/main/webapp/app/account/register/register.component.html
Comment (krusche) [reply src/main/webapp/app/account/register/register.component.html:12]: It really helped me in testing the regex pattern, so this might be super useful for other administrators playing around with Artemis, therefore I would like to keep it.

File: src/main/webapp/app/account/register/register.component.ts
Comment (krusche) [reply src/main/webapp/app/account/register/register.component.ts:54]: It really helped me in testing the regex pattern, so this might be super useful for other administrators playing around with Artemis, therefore I would like to keep it.

**Review by kloessst — DISMISSED** [review]
Short answer quiz statistics work now:

image

Registration and email regex worked.
E-Mail activation works as expected now. Had to activate before I could log in.

**Review by sleiss — APPROVED** [review]
Re-Approve.

File: src/main/webapp/app/account/settings/settings.component.ts
Comment (MaisiKoleni) [inline src/main/webapp/app/account/settings/settings.component.ts:21]: This is another maximum email length

File: src/test/resources/config/application-artemis.yml
Comment (MaisiKoleni) [inline src/test/resources/config/application-artemis.yml:26]:
```suggestion
            allowed-email-pattern: [a-zA-Z0-9_\-\.\+]+@[a-zA-Z0-9_\-\.]+\.[a-zA-Z]{2,5}
```

**Review by MaisiKoleni — COMMENTED** [review]
Found some small issues again, I think most of them were there before and could be done in another PR.

Tested the registration on TS2 and it worked fine.

**Comment by MaisiKoleni** [comment]
I am now done with reviewing most of the 159 files. If I should take a look at some of them specifically again, let me know.

File: src/main/webapp/i18n/de/global.json
Comment (krusche) [reply src/main/webapp/i18n/de/global.json:139]: thx. I fixed it.

File: src/main/webapp/i18n/de/global.json
Comment (krusche) [reply src/main/webapp/i18n/de/global.json:145]: thx. I fixed it.

File: src/main/webapp/i18n/en/global.json
Comment (krusche) [reply src/main/webapp/i18n/en/global.json:145]: thx. I fixed it.

File: src/main/webapp/app/account/register/register.component.html
Comment (krusche) [reply src/main/webapp/app/account/register/register.component.html:121]: thx. I fixed it.

File: src/main/webapp/app/account/settings/settings.component.ts
Comment (krusche) [reply src/main/webapp/app/account/settings/settings.component.ts]: thx. I fixed it.

File: src/test/resources/config/application-artemis.yml
Comment (krusche) [reply src/test/resources/config/application-artemis.yml:26]: if we write it like this, we have to put it into `'...'`

**Comment by krusche** [comment]
> I am now done with reviewing most of the 159 files. If I should take a look at some of them specifically again, let me know.

Thanks. I fixed the mentioned issues. I guess we are fine for now. I'll wait for the build and then I'll merge

**Review by MaisiKoleni — APPROVED** [review]

**Review by sleiss — APPROVED** [review]

**Comment by nk-coding** [comment]
> Thanks. Improved error messages will be implemented in a follow-up

Are there any plans to implement this?  
It happened several times that someone tried to use an unavailable username and therefore could not register.  
Better error messages would help a lot.

**Pull request merged**
